### DISCUSSION TITLE
Extension framework enhancements.

### DISCUSSION THREAD

**Discussion by singularcitrus** [Paymenter/Paymenter#369] [Ideas]
Hi, I was trying to integrate my local payment gateway and found a few issues. For the record, here is the payment gateway docs because I will be referencing concepts from here https://developers.payfast.co.za/docs

- Allow plugins to redirect with post data (similar to users clicking "submit" on a form), this is how this payment gateway receives its information.
- Have the ability for plugins to store data, like tokens. https://developers.payfast.co.za/docs#tokenization
- Add a headless pay function or something similar that can be used in automatic billing. This will also help with the subscriptions that were suggested in #254

**Comment by singularcitrus** [comment]
Adding to this, you would also then want functions to update, pause, and cancel subscriptions/tokens

**Comment by CorwinDev** [comment]
> Allow plugins to redirect with post data (similar to users clicking "submit" on a form), this is how this payment gateway receives its information.

This can already be done, take a look at our [extensions](https://github.com/Paymenter/Paymenter/blob/master/app/Extensions/Gateways/PayPal/PayPal.php#L40-L73)

> Have the ability for plugins to store data, like tokens. https://developers.payfast.co.za/docs#tokenization

This can also be done, take a look at [Stripe Subscriptions](https://github.com/Paymenter/Paymenter/blob/master/app/Extensions/Gateways/StripeSubscriptions/StripeSubscriptions.php#L216)

**Comment by singularcitrus** [comment]
> This can already be done, take a look at our [extensions](https://github.com/Paymenter/Paymenter/blob/master/app/Extensions/Gateways/PayPal/PayPal.php#L40-L73)

I see what you mean, I did try that yesterday. I used the paypal extension as a template and it didn't work. I'll try again, maybe the error I got was unrelated and prevented execution of something.

> This can also be done, take a look at [Stripe Subscriptions](https://github.com/Paymenter/Paymenter/blob/master/app/Extensions/Gateways/StripeSubscriptions/StripeSubscriptions.php#L216)

Aha, this is great, thanks! I did not see it.


The other ideas I have is still important though, since ideally you don't want to bother users with invoices every month and automatically bill invoices.

**Comment by singularcitrus** [comment]
> This can already be done, take a look at our [extensions](https://github.com/Paymenter/Paymenter/blob/master/app/Extensions/Gateways/PayPal/PayPal.php#L40-L73)

@CorwinDev I have once again tried it and it does not work. When you post to the process payment page of Payfast it returns html and not json. You have to _redirect_ to the page _with post data_, and not simply post to it, here is the code I used:
```php
$response = Http::withHeaders([
    'Content-Type' => 'application/json',
])->post("https://sandbox.payfast.co.za/eng/process", $data);

\Log::info($response);
```
In this case `$response` is a string of HTML.

Edit: I have also tried with ...`->asForm()`... and same result

Another edit:
At the moment I have a different function that renders a blank page with a "Pay Now" button that is the form as described in the Payfast docs:
```php
$htmlForm = '<form action="https://'.$pfHost.'/eng/process" method="post">';
foreach($data as $name=> $value)
{
    $htmlForm .= '<input name="'.$name.'" type="hidden" value=\''.$value.'\' />';
}
$htmlForm .= '<input id="pay" type="submit" value="Pay Now"/></form>';
echo $htmlForm;
```

**Reply by CorwinDev** [reply]
https://github.com/Paymenter/Extensions/blob/v0.8/Gateways/PayU/PayU.php#L91-L97

The payu code may help a bit with this :D
